client/my-sites/checkout/checkout-thank-you/marketplace/marketplace-thank-you.tsx

```diff
@@ -114,6 +121,48 @@ const MarketplaceThankYou = ( { productSlug }: { productSlug: string } ) => {
 		// eslint-disable-next-line react-hooks/exhaustive-deps
 	}, [] );
 
+	// Set progressbar (currentStep) depending on transfer.status and pluginOnSite.
+	// Step 0 hides the progress bar. It means "complete" or "still loading transfer status".
+	// Steps 1-4 advance through the bar.
+	useEffect( () => {
+		if ( transfer?.status === AtomicTransferActive ) {
+			setCurrentStep( 1 );
+		} else if ( transfer?.status === AtomicTransferProvisioned ) {
+			setCurrentStep( 2 );
+		} else if ( transfer?.status === AtomicTransferRelocating ) {
+			setCurrentStep( 3 );
+		} else if ( transfer?.status === AtomicTransferComplete ) {
+			/*
+			 * When we're done transferring, we want to see pluginOnSite to advance
+			 * to step 0 (complete).
+			 * Even after pluginOnSite exists, subsequent requests temporarily set
+			 * pluginOnSite = undefined when there are outstanding network requests.
+			 * Workaround: After seeing pluginOnSite once, we cannot go back to step 4.
+			 */
+			if ( ! pluginOnSite && ! sawPluginOnce ) {
+				setCurrentStep( 4 );
+			} else if ( pluginOnSite && ! sawPluginOnce ) {
+				setCurrentStep( 0 );
+				setSawPluginOnce( true );
+			} else {
+				setCurrentStep( 0 );
+			}
+		} else {
+			setCurrentStep( 0 );
+		}
+	}, [ transfer, pluginOnSite, sawPluginOnce ] );
```

Comment:
It's possible that the `transfer` object will be recreated on more renders than you intend. Since you don't actually need it in this effect, I suggest instead depending on `transfer.status` since that is a string and will be compared by value rather than by reference.

```js
const transferStatus = transfer?.status;
useEffect(() => {
  if ( transferStatus === AtomicTransferActive ) {
    // ...
}, [ transferStatus, pluginOnSite, sawPluginOnce ] );
```

You might also want to make sure that `pluginOnSite` does not change regularly. Following the selectors it looks like it might. Since you only need to know if it is set, you could create a boolean (which again will be compared by value) instead of depending on the object.

```
const isPluginOnSite = !! pluginOnSite;
```

Reply:
Good find, thank you!
Some of the other instances of useEffect are intentional, to create repeated polling, but I found at least one that could be updated.